libyul/optimiser/ForLoopConditionIntoBody.cpp

```diff
@@ -54,3 +56,45 @@ void ForLoopConditionIntoBody::operator()(ForLoop& _forLoop)
 	ASTModifier::operator()(_forLoop);
 }
 
+void ForLoopConditionOutOfBody::operator()(ForLoop& _forLoop)
+{
+	ASTModifier::operator()(_forLoop);
+
+	if (
+		!m_dialect.booleanNegationFunction() ||
+		_forLoop.condition->type() != typeid(Literal) ||
+		valueOfLiteral(boost::get<Literal>(*_forLoop.condition)) == u256(0) ||
+		_forLoop.body.statements.empty() ||
+		_forLoop.body.statements.front().type() != typeid(If)
+	)
+		return;
+
+	If& firstStatement = boost::get<If>(_forLoop.body.statements.front());
+	if (
+		firstStatement.body.statements.size() != 1 ||
+		firstStatement.body.statements.front().type() != typeid(Break)
+	)
+		return;
+	if (!SideEffectsCollector(m_dialect, *firstStatement.condition).movable())
+		return;
+
+	YulString iszero = m_dialect.booleanNegationFunction()->name;
+	langutil::SourceLocation location = locationOf(*firstStatement.condition);
+
+	if (
+		firstStatement.condition->type() == typeid(FunctionCall) &&
+		boost::get<FunctionCall>(*firstStatement.condition).functionName.name == iszero
+	)
+		_forLoop.condition = make_unique<Expression>(std::move(boost::get<FunctionCall>(*firstStatement.condition).arguments.front()));
+	else
+		_forLoop.condition = make_unique<Expression>(FunctionCall{
```

Comment:
We could always do this - ``iszero(iszero(x))`` -> ``x`` should be a simplified for ``if`` conditions elsewhere, right?